internal/runtime/internal/controller/value_cache.go

```diff
@@ -234,3 +245,49 @@ func (vc *valueCache) buildValue(from []ComponentID, offset int) interface{} {
 	}
 	return attrs
 }
+
+// mergeScopes merges the Variables maps of the scope hierarchy into a new map.
+// The child always overrides the parent's value.
+func mergeScopes(scope *vm.Scope) map[string]any {
```

Comment:
Do you think there is any performance impact from this? I suppose it depends on the scenario, since there might be cases where it's efficient to merge upfront e.g. when there's a deeply nexted structure of modules and we instantiate a lot components from the outermost layer inside the innermost layer.

Reply:
Without the parent, the merge is a bit simpler but still needed. Both solutions have a performance impact though because the scope is never nil (it contains at least the [module_path](https://github.com/grafana/alloy/blob/main/internal/runtime/alloy.go#L312))

Reply:
Maybe it's time for a deeper refactor of the value_cache. I will try something different.